// cdktf/lib/validations/validate-binary-version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateBinaryVersion = void 0;
const child_process_1 = require("child_process");
const semver = require("semver");
/**
 * A validation that can be applied to a construct that will error if the
 * construct is being used in an environment with a version of a binary lower than the specified version.
 */
class ValidateBinaryVersion {
    constructor(binary, versionConstraint, versionCommand, hint) {
        this.binary = binary;
        this.versionConstraint = versionConstraint;
        this.versionCommand = versionCommand;
        this.hint = hint;
    }
    validate() {
        try {
            const versionOfCommand = (0, child_process_1.execSync)(this.versionCommand, {
                stdio: "pipe",
            }).toString();
            const semverRegex = /\d+\.\d+\.\d+/;
            const version = versionOfCommand.match(semverRegex);
            if (!version) {
                return [
                    `Could not determine version of ${this.binary} (running ${this.versionCommand})`,
                ];
            }
            if (!semver.satisfies(version[0], this.versionConstraint)) {
                return [
                    `${this.binary} version ${version[0]} is lower than the required version ${this.versionConstraint} for this construct. ${this.hint || ""}`,
                ];
            }
            return [];
        }
        catch (e) {
            return [
                `Could not determine version of ${this.binary}, ${this.versionCommand} failed: ${e}`,
            ];
        }
    }
}
exports.ValidateBinaryVersion = ValidateBinaryVersion;
